File: 백준/Gold/6443. 애너그램/애너그램.py
n = int(input())


def dfs(energram_dict, answer):
    if len(answer) == len(a):
        print("".join(answer))
        return

    for energram in energram_dict:
        if energram_dict[energram]:
            energram_dict[energram] -= 1
            answer.append(energram)
            dfs(energram_dict, answer)
            answer.pop()
            energram_dict[energram] += 1
    return


for _ in range(n):
    a = list(input())
    a.sort()
    visited = [0]*len(a)
    energram_dict = dict()

    for i in a:
        if i in energram_dict.keys():
            energram_dict[i] += 1
        else:
            energram_dict[i] = 1


    dfs(energram_dict, [])

File: 백준/Gold/6443. 애너그램/애너그램.js
const readline = require("readline");

// 입출력을 위한 인터페이스 객체 생성
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];

rl.on("line", (line) => {
  input.push(line);
  if (input.length == Number(input[0]) + 1) {
    rl.close();
  }
}).on("close", () => {
  const n = Number(input[0]);
  const answer = new Set();

  for (let i = 1; i <= n; i++) {
    const arr = input[i].split("");
    arr.sort();
    const word = [];
    let dict = {};

    for (let a of arr) {
      if (a in dict) {
        dict[a] += 1;
      } else {
        dict[a] = 1;
      }
    }
    const dfs = (depth, word) => {
      if (depth === arr.length) {
        answer.add(word.join(""));
        return;
      }
      for (let key in dict) {
        if (dict[key]) {
          dict[key] -= 1;
          word.push(key);
          dfs(depth + 1, word);
          word.pop();
          dict[key] += 1;
        }
      }
      return;
    };

    dfs(0, word);
  }
  const answer_arr = [...answer];
  console.log(answer_arr.join("\n"));
  process.exit();
});
